// @mikro-orm/migrations/dist/MigrationRunner.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MigrationRunner = void 0;
const core_1 = require("@mikro-orm/core");
class MigrationRunner {
    constructor(driver, options, config) {
        this.driver = driver;
        this.options = options;
        this.config = config;
        this.connection = this.driver.getConnection();
        this.helper = this.driver.getPlatform().getSchemaHelper();
    }
    async run(migration, method) {
        migration.reset();
        await migration[method]();
        let queries = migration.getQueries();
        if (this.options.disableForeignKeys) {
            const charset = this.config.get('charset');
            queries.unshift(...this.helper.getSchemaBeginning(charset).split('\n'));
            queries.push(...this.helper.getSchemaEnd().split('\n'));
        }
        queries = queries.filter(sql => sql.trim().length > 0);
        if (!this.options.transactional || !migration.isTransactional()) {
            await core_1.Utils.runSerial(queries, sql => this.driver.execute(sql));
            return;
        }
        await this.connection.transactional(async (tx) => {
            await core_1.Utils.runSerial(queries, sql => this.driver.execute(tx.raw(sql)));
        }, this.masterTransaction);
    }
    setMasterMigration(trx) {
        this.masterTransaction = trx;
    }
    unsetMasterMigration() {
        delete this.masterTransaction;
    }
}
exports.MigrationRunner = MigrationRunner;
